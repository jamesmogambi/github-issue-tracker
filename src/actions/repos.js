require("isomorphic-fetch");

// SET_REPOSITORIES
export const setRepositories = (payload) => ({
  type: "SET_REPOSITORIES",
  payload,
});
// CLEAR_REPOSITORIES
export const clearRepositories = () => ({
  type: "CLEAR_REPOSITORIES",
});

// ERROR
export const error = (payload) => ({
  type: "ERROR",
  payload,
});

// Get Github Repositories
export const getRepositories = (params) => {
  return (dispatch) => {
    dispatch(clearRepositories());
    fetch("https://api.github.com/graphql", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        authorization: `bearer ${process.env.GITHUB_ACCESS_TOKEN}`,
      },
      body: JSON.stringify({
        query: `
               query($github_login: String!) {
    user(login: $github_login) {
      name
      repositories(first: 100) { 
        nodes {
             id
        name
        description 
        url 
        updatedAt
        forkCount
        }
      }
    }
  }
                 `,
        variables: {
          github_login: params.username,
        },
      }),
    })
      .then((res) => res.json())
      .then((res) => {
        if (res.data.user) {
          let repos = res.data.user.repositories.nodes;
          let owner = res.data.user.name;
          dispatch(
            setRepositories({
              repositories: repos,
              repositoryOwner: params.username,
            })
          );
        }
      })
      .catch((err) => {
        dispatch(error(err));
        console.error(err);
      });
  };
};
