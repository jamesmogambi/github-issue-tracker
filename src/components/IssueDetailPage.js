import React from "react";
import { connect } from "react-redux";

export const IssueDetailsPage = (props) => (
  <div>
    <div className="content-container">
      {props.issue ? (
        <div>
          <h1 className="page-header__title">
            Viewing details of Issue <span>{props.issue.node.title}</span>{" "}
          </h1>
          <div
            dangerouslySetInnerHTML={{ __html: props.issue.node.bodyHTML }}
          ></div>
        </div>
      ) : (
        <div className="list-item list-item--message">
          <span>No Details </span>
        </div>
      )}
    </div>
  </div>
);

const mapStateToProps = (state, props) => {
  return {
    issue: state.issues.repositoryissues.find(
      (issue) => issue.node.id === props.match.params.id
    ),
  };
};

export default connect(mapStateToProps)(IssueDetailsPage);
