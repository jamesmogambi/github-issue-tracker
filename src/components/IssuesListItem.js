import React from "react";
import { Link } from "react-router-dom";

const IssuesListItem = ({ id, title, author, createdAt }) => (
  <Link className="list-item" to={`/issueDetails/${id}`}>
    <div>
      <img src={author.avatarUrl} alt="Avatar" className="avatar" />
      <div className="list-item__sub-title">{title}</div>
    </div>
    <span className="list-item__sub-title ">opened at {createdAt}</span>
  </Link>
);

export default IssuesListItem;
