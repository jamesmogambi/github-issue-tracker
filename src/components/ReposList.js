import React from "react";
import { connect } from "react-redux";
import ReposListItem from "./ReposListItem";

export const ReposList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Repositories</div>
      <div className="show-for-desktop">Repository</div>
      <div className="show-for-desktop">Action</div>
    </div>
    <div className="list-body">
      {props.repos.repositories.length === 0 ? (
        <div className="list-item list-item--message">
          <span>No Repos</span>
        </div>
      ) : (
        props.repos.repositories.map((repo) => {
          return (
            <ReposListItem
              key={repo.id}
              repositoryOwner={props.repos.repositoryOwner}
              {...repo}
            />
          );
        })
      )}
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    repos: state.repos,
  };
};

export default connect(mapStateToProps)(ReposList);
