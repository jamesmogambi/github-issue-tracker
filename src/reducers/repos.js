// Repositories Reducer
const initialState = {
  repositories: [],
  repositoryOwner: null,
  error: null,
};

export default (state = initialState, action) => {
  switch (action.type) {
    case "SET_REPOSITORIES":
      const { repositories, repositoryOwner } = action.payload;
      return {
        ...state,
        repositoryOwner,
        repositories,
      };
    case "CLEAR_REPOSITORIES":
      return {
        repositories: [],
        repositoryOwner: null,
        error: null,
      };

    default:
      return state;
  }
};
