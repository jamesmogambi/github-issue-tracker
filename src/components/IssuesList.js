import React from "react";
import { connect } from "react-redux";
import { getRepositoryIssues } from "../actions/issues";
import IssuesListItem from "./IssuesListItem";
import selectIssues from "../selectors/issues";

// export class IssuesList extends React.Component {
//   constructor(props) {
//     super(props);
//   }
//   componentWillMount() {
//     this.props.getRepositoryIssues(this.props.params);
//   }
//   render() {
//     return (
//       <div className="content-container">
//         <h1 className="page-header__title">
//           Viewing Issues of Repository <span>{this.props.params.repoName}</span>{" "}
//         </h1>

//         <div className="list-header">
//           <div className="show-for-mobile">Issues</div>
//           <div className="show-for-desktop">Issue</div>
//           <div className="show-for-desktop">Created</div>
//         </div>
//         <div className="list-body">
//           {this.props.issues.length === 0 ? (
//             <div className="list-item list-item--message">
//               <span>No Issues for this Repo</span>
//             </div>
//           ) : (
//             this.props.issues.map(({ node }) => {
//               return <IssuesListItem key={node.id} {...node} />;
//             })
//           )}
//         </div>
//       </div>
//     );
//   }
// }

// const mapDispatchToProps = (dispatch) => ({
//   getRepositoryIssues: (params) => dispatch(getRepositoryIssues(params)),
// });

// const mapStateToProps = (state) => {
//   return {
//     issues: selectIssues(state.issues.repositoryissues, state.filters),
//   };
// };

// export default connect(mapStateToProps, mapDispatchToProps)(IssuesList);

export const IssuesList = (props) => (
  <div className="content-container">
    <h1 className="page-header__title">
      Viewing Issues of Repository <span>{props.repoName}</span>{" "}
    </h1>

    <div className="list-header">
      <div className="show-for-mobile">Issues</div>
      <div className="show-for-desktop">Issue</div>
      <div className="show-for-desktop">Created</div>
    </div>
    <div className="list-body">
      {props.issues.length === 0 ? (
        <div className="list-item list-item--message">
          <span>No Issues for this Repo</span>
        </div>
      ) : (
        props.issues.map(({ node }) => {
          return <IssuesListItem key={node.id} {...node} />;
        })
      )}
    </div>
  </div>
);

export default IssuesList;
