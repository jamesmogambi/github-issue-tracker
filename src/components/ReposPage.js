import React from "react";
import { connect } from "react-redux";
import Header from "./Header";
import UserForm from "./UserForm";
import ReposList from "./ReposList";
import { getRepositories } from "../actions/repos";

export class ReposPage extends React.Component {
  onSubmit = (params) => {
    this.props.getRepositories(params);
  };
  render() {
    return (
      <div>
        <div className="content-container">
          <UserForm onSubmit={this.onSubmit} />
          <ReposList />
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  getRepositories: (params) => dispatch(getRepositories(params)),
});

export default connect(undefined, mapDispatchToProps)(ReposPage);
