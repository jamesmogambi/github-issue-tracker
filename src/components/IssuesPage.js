import React from "react";
import { connect } from "react-redux";
import { getRepositoryIssues } from "../actions/issues";
import IssuesList from "./IssuesList";
import IssuesListFilters from "./IssuesListFilters";
import selectIssues from "../selectors/issues";

export class IssuesPage extends React.Component {
  constructor(props) {
    super(props);
  }
  componentWillMount() {
    this.props.getRepositoryIssues(this.props.match.params);
  }
  render() {
    return (
      <div>
        <IssuesListFilters />
        <IssuesList issues={this.props.issues} repoName={this.props.repoName} />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  getRepositoryIssues: (params) => dispatch(getRepositoryIssues(params)),
});

const mapStateToProps = (state) => {
  return {
    issues: selectIssues(state.issues.repositoryissues, state.filters),
    repoName: state.issues.repositoryname,
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(IssuesPage);
