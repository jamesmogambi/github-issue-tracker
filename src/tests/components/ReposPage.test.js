import React from "react";
import { shallow } from "enzyme";
import repos from "../fixtures/repos";
import { ReposPage } from "../../components/ReposPage";

let getRepositories, wrapper;

beforeEach(() => {
  getRepositories = jest.fn();
  wrapper = shallow(<ReposPage getRepositories={getRepositories} />);
});

test("should render ReposPage", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should handle getRepositories", () => {
  const params = {
    username: "JamesEdward",
  };
  wrapper.find("UserForm").prop("onSubmit")(params);
  expect(getRepositories).toHaveBeenLastCalledWith(params);
});
