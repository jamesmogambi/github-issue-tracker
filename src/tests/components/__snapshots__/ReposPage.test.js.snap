// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ReposPage 1`] = `
<div>
  <div
    className="content-container"
  >
    <UserForm
      onSubmit={[Function]}
    />
    <Connect(ReposList) />
  </div>
</div>
`;
